asmcomp/riscv/emit.mlp

```diff
@@ -302,20 +311,45 @@ let emit_instr env i =
         emit_stack_adjustment n;
         `	{emit_tail func}\n`
       end
-  | Lop(Iextcall{func; alloc = true}) ->
-      `	la	{emit_reg reg_t2}, {emit_symbol func}\n`;
-      `	{emit_call "caml_c_call"}\n`;
-      record_frame env i.live (Dbg_other i.dbg)
-  | Lop(Iextcall{func; alloc = false}) ->
-      `	{emit_call func}\n`
+  | Lop(Iextcall{func; alloc; stack_ofs}) ->
+      if stack_ofs > 0 then begin
+        `	mv	{emit_reg reg_stack_arg_begin}, sp\n`;
+        `	addi	{emit_reg reg_stack_arg_end}, sp, {emit_int (Misc.align stack_ofs 16)}\n`;
+        `	la	{emit_reg reg_t2}, {emit_symbol func}\n`;
+        `	{emit_call "caml_c_call_stack_args"}\n`;
+        record_frame env i.live (Dbg_other i.dbg)
+      end else if alloc then begin
+        `	la	{emit_reg reg_t2}, {emit_symbol func}\n`;
+        `	{emit_call "caml_c_call"}\n`;
+        record_frame env i.live (Dbg_other i.dbg)
+      end else begin
+        (* store ocaml stack in s0, which is marked as being destroyed
+           at noalloc calls *)
+        `	mv	s0, sp\n`;
+        cfi_remember_state ();
+        cfi_def_cfa_register ~reg:21;
+        let ofs = Domainstate.(idx_of_field Domain_c_stack) * 8 in
+        `	ld	sp, {emit_int ofs}({emit_reg reg_domain_state_ptr})\n`;
+        `	{emit_call func}\n`;
+        `	mv	sp, s0\n`;
+        cfi_restore_state ()
+      end
   | Lop(Istackoffset n) ->
       assert (n mod 16 = 0);
       emit_stack_adjustment (-n);
-      adjust_stack_offset env n
-  | Lop(Iload { memory_chunk = Single; addressing_mode = Iindexed ofs; _ } ) ->
+      env.stack_offset <- env.stack_offset + n
+  | Lop(Iload { memory_chunk = Single; addressing_mode = Iindexed ofs; is_atomic } ) ->
+      assert (not is_atomic);
       `	flw	{emit_reg i.res.(0)}, {emit_int ofs}({emit_reg i.arg.(0)})\n`;
       `	fcvt.d.s	{emit_reg i.res.(0)}, {emit_reg i.res.(0)}\n`
-  | Lop(Iload { memory_chunk; addressing_mode = Iindexed ofs } ) ->
+  | Lop(Iload { memory_chunk = Word_int | Word_val; addressing_mode = Iindexed ofs; is_atomic } ) ->
+      if is_atomic then
+        `	fence rw, rw\n`;
+      `	ld	{emit_reg i.res.(0)}, {emit_int ofs}({emit_reg i.arg.(0)})\n`;
```

Comment:
nit: alignment.

Reply:
You mean to use a TAB instead of SPACE between `fence` and `rw`? Fixed!

Reply:
I was pointing out that the backticks are not left-aligned.

Reply:
Looking at the existing backend code, it seems that the backticks are aligned as normal OCaml code, so indented under an `if`. But I agree it looks a bit odd. Instead I put it on the same line as the `if`, similar to https://github.com/ocaml/ocaml/blob/31ba6edb0458e2b377d674257d9ccbd2c1e82804/asmcomp/arm64/emit.mlp#L816

Reply:
Thanks. I was reading the code wrong 😑 . The fix looks good.